### ISSUE TITLE
Improve speed of `density.ppp()` with `at = "pixels"`

### ISSUE THREAD

**Issue by edwardlavender** [spatstat/spatstat.explore#4]
Thank you for an amazing set of resources and packages for spatial point patterns. 

This is a feature request. I am wondering if you can improve the speed of `density.ppp()` when `at = "pixels"` is used. This is fast for small grids but with large grids it becomes agonisingly slow. If anything can be done to speed this up under the hood, it would be extremely helpful.

**Comment by baddstats** [comment]
Please give an example where you consider the speed is unacceptably slow.

**Comment by edwardlavender** [comment]
Thank you for your response. Below is a simple reprex. We are working on patterns of space use for aquatic animals in coastal areas. In the example below, I download some coastline data for a relatively small area off the west coast of Scotland and I sample a limited number of points ('animal locations') in that area. With the default options, the function runs in < 5 s on my computer. But even in this example we need ~500 pixels to get a reasonable map, given the complex coastline, which takes ~1 minute with the default sigma. In reality, we are working with many more points so computation time is often much slower  (and we are often working with data from numerous simulations, or many separate individuals, so computation times add up quickly). Anything that can be done to boost speed would be greatly appreciated. 

```
library(spatstat)
library(terra)
library(sf)
library(tictoc)

set.seed(123)

#### Step 1: Define study area (observation window)
# Define boundaries of study area (UTM 29N)
bb <- c(640000, 739999, 6179749, 6318933)
# Download land boundaries & convert to UTM 29N
land <- geodata::gadm(country = 'United Kingdom', path = tempdir())
land <- terra::project(land, "EPSG:32629")
land <- terra::crop(land, terra::ext(bb))
# Define polygon of sea (for our observation window)
land <- 
  land |>
  st_as_sf() |> 
  st_union()
sea <- 
  land |> 
  st_bbox() |>
  st_as_sfc() |>
  st_difference(land) |> 
  terra::vect()

#### Step 2: Define example points
# Sample some points in the sea
xy <- 
  terra::spatSample(sea, size = 200L) |> 
  terra::crds() |> 
  as.data.frame()
# (optional) Visualise study area & points
terra::plot(sea)
points(xy$x, xy$y, col = "red")

#### Step 3: Prepare ppp object
win <- as.owin(st_as_sf(sea))
X   <- ppp(x = xy$x, y = xy$y, window = win)
# plot(X)

#### Step 4: Estimate density with default no. pixels
spatstat.options("npixel" = 128)
tic()
D <- density.ppp(X, at = "pixels")
toc()
# 3.934 sec elapsed
plot(D)

#### Step 5: Estimate density with more pixels 
# (Otherwise, use default options for speed)
spatstat.options("npixel" = 500)
tic()
D <- density.ppp(X, at = "pixels")
toc()
# 56.632 sec elapsed
plot(D)
```

**Comment by baddstats** [comment]
OK thanks for the example.

Most of the computation time in this example is taken up by the discretisation of the polygonal window onto the pixel raster (which is needed for the edge correction calculation). The actual density calculations (using the Fast Fourier Transform) take only a fraction of a second. 

To make this go faster you could 
- simplify the polygon (e.g. `Window(X) <- simplify.owin(Window(X), dmin)`)
- switch off edge correction by calling `density.ppp` with `edge=FALSE` (not recommended)
- discretise the window first by  `X <- discretise(X, dimyx=500)` which will take approx 50 sec but will allow you to try different smoothing parameters without repeating the discretisation step

But ... 56 seconds is agonisingly slow?!? I guess I come from an older generation %^]

**Comment by edwardlavender** [comment]
Thank you very much for the really quick response! This is much appreciated & very helpful.

Regarding computation time, I chose my words poorly! The computation time here (56 s) is not a big deal (though this is just one example). With 200 points and `sigma = bw.diggle`, computation time goes up to 50 s (128 pixels) and 234 s (500 pixels) on my machine. In real-world analyses, I have experienced longer times (probably because our polygons are more complex than in the above example). Individually, times on the order of several minutes are not a big deal, but they do add up--so these tips to reduce computation time are extremely valuable.  

In my use cases so far, the advice above largely resolve this issue. Simplifying the polygon in particular makes an enormous improvement to computation time. I'll let you know if we run into any other examples where computation time is slow. Thanks again!
